### PR TITLE
Resolve 2 spotbugs issues and increase spotbugs checks

### PR DESCRIPTION
**Pull request by prince-panwar**
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did

### CODE REVIEW

**Comment by prince-panwar** [comment]
hi, mark I have successfully fixed 3 out of 5 bugs , but I am stuck at two similar bugs:-
In the code, both the gitLabAPI and me fields are marked as transient, which means they will not be persisted when the object is serialized. This is leading to the error  "SE_TRANSIENT_FIELD_NOT_RESTORED" indicates that a field marked as transient has not been restored properly after being deserialized.
can you help me out here?

**Comment by prince-panwar** [comment]
I am not sure what is the best way to deal with these bugs

File: src/main/java/org/jenkinsci/plugins/GitLabAuthenticationToken.java [hunk 1]
```diff
@@ -72,6 +72,12 @@ public class GitLabAuthenticationToken extends AbstractAuthenticationToken {
     private final transient User me;
     private transient GitLabSecurityRealm myRealm = null;
 
+
+
+
+
+
+
```
Comment (MarkEWaite) [inline src/main/java/org/jenkinsci/plugins/GitLabAuthenticationToken.java:80]: I'd remove this change, since it does not seem relevant to the pull request
```suggestion
```

**Comment by MarkEWaite** [comment]
> can you help me out here?

Add an exclusions filter file as [described in the tutorial](https://www.jenkins.io/doc/developer/tutorial-improve/add-more-spotbugs-checks/#add-a-spotbugs-exclusions-file-if-needed).  Exclude those two issues.

It is also OK to include all 5 issues in the exclusions filter file.  Excluding a spotbugs warning does not make the code any worse and reduces the risk that future changes will introduce issues that could have been detected by spotbugs.

File: src/main/java/org/jenkinsci/plugins/GitLabOAuthUserDetails.java [hunk 2]
```diff
@@ -6,21 +6,18 @@
...
 import org.acegisecurity.userdetails.User;
-import org.acegisecurity.userdetails.UserDetails;
+
 
 /**
  * @author Mike
  *
  */
-public class GitLabOAuthUserDetails extends User implements UserDetails {
+public class GitLabOAuthUserDetails extends User {
```
Comment (MarkEWaite) [inline src/main/java/org/jenkinsci/plugins/GitLabOAuthUserDetails.java:15]: This seems like a breaking change.  Why was the `implements UserDetails` removed from the class?

File: src/main/java/org/jenkinsci/plugins/GitLabSecurityRealm.java [hunk 3]
```diff
@@ -221,24 +220,25 @@ public Object unmarshal(HierarchicalStreamReader reader, UnmarshallingContext co
...
         private void setValue(GitLabSecurityRealm realm, String node, String value) {
-            switch (node.toLowerCase()) {
-                case "clientid":
-                    realm.setClientID(value);
-                    break;
-                case "clientsecret":
-                    realm.setClientSecret(value);
-                    break;
-                case "gitlabweburi":
+            switch (node) {
```
Comment (MarkEWaite) [inline src/main/java/org/jenkinsci/plugins/GitLabSecurityRealm.java:223]: Since all the cases that are being tested are English language strings, why not set the locale to English in the `node.toLowerCase()` call rather than risk that some caller was passing something that differed only by the case of the characters?

File: src/main/java/org/jenkinsci/plugins/GitLabSecurityRealm.java [hunk 4]
```diff
@@ -221,24 +220,25 @@ public Object unmarshal(HierarchicalStreamReader reader, UnmarshallingContext co
...
                     realm.setGitlabApiUri(value);
                     break;
+                case "clientID":
+                    realm.setClientID(value);
+                    break;
+                case "clientSecret":
+                    realm.setClientSecret(value);
+                    break;
                 default:
-                    throw new ConversionException("Invalid node value = " + node);
```
Comment (MarkEWaite) [inline src/main/java/org/jenkinsci/plugins/GitLabSecurityRealm.java:238]: Why was the exception removed?

**Review by MarkEWaite — CHANGES_REQUESTED** [review]
Thanks!  I think that some of the changes may be safer to exclude in the filter file rather than make a change in the production code.  Alternate ideas are welcome.  See my questions with individual changes.

**Comment by prince-panwar** [comment]
thanks, I'll fix these in next commit

**Comment by prince-panwar** [comment]
I added the exclusion file but spot bugs is still showing bugs . what am i doing wrong?

**Comment by MarkEWaite** [comment]
> I added the exclusion file but spot bugs is still showing bugs . what am i doing wrong?

There were several mistakes in `src/spotbugs/excludesFilter.xml`.  I've updated the file to correct those mistakes in e42bc192edf2b5d2ccc518bf7ea84d66034a3f80

Mistakes included:

* Used `Class` when you also needed to use `Field`.  I switched it to use both `Class` and `Field`
* Used `Class` when you also nee to use `Method`.  I switched it to use both `Class` and `Method`
* Used more complicated `And` and `Or` when declaring multiple `Match` conditions is easier to read.  I switched to multiple top level `Match` conditions

Thanks for the submission!

**Comment by MarkEWaite** [comment]
I also set the `checkstyle` version in 0b32055a5d79e1888372bad461e1617a61d86d11 in order to prevent a maven warning at compile time.  I'm not sure why that wasn't detected previously, but it is resolved now.

**Comment by MarkEWaite** [comment]
I needed b7510e77a6d86dd31f40d3a5900206b04b203c98 to undo the change to the `setValue` method otherwise it would have entered the `switch` with a lowercase value and then been checking strings that contain uppercase letters.  That would have failed.

File: src/main/java/org/jenkinsci/plugins/GitLabOAuthUserDetails.java (hunk 2 above)
Comment (MarkEWaite) [reply src/main/java/org/jenkinsci/plugins/GitLabOAuthUserDetails.java:15]: Thanks!  I'd never seen that spotbugs warning before.  There appears to be no end to the things that I need to lear!

File: src/main/java/org/jenkinsci/plugins/GitLabOAuthUserDetails.java [hunk 5]
```diff
@@ -6,21 +6,18 @@
...
 
-    /**
-     *
-     */
     private static final long serialVersionUID = 1709511212188366292L;
 
     public GitLabOAuthUserDetails(org.gitlab4j.api.models.User user, GrantedAuthority[] authorities) {
-        super(user.getUsername(), "", true, true, true, true, authorities);
+        super(user.getUsername(), user.getEmail(), true, true, true, true, authorities);
```
Comment (MarkEWaite) [inline src/main/java/org/jenkinsci/plugins/GitLabOAuthUserDetails.java:20]: Why was this changed?  I don't see any indication in the spotbugs output that it needs to be changed and I see no indication in any of the commit messages why it needs to be changed.  What did I miss?
```suggestion
        super(user.getUsername(), "", true, true, true, true, authorities);
```

Comment (MarkEWaite) [reply src/main/java/org/jenkinsci/plugins/GitLabOAuthUserDetails.java]: I reverted that change in b8e5404eb11b22fbe8c00647038f3ed89f849c20 .  If there is a reason to restore that change, please do so with an additional commit that includes a message describing the reason for the change.

**Review by MarkEWaite — APPROVED** [review]
Approved with changes as they are now.

**Comment by MarkEWaite** [comment]
Should be squash merged when it is merged.

File: src/main/java/org/jenkinsci/plugins/GitLabOAuthUserDetails.java (hunk 5 above)
Comment (prince-panwar) [reply src/main/java/org/jenkinsci/plugins/GitLabOAuthUserDetails.java]: hard-coded values for the User constructor parameters might not be appropriate for all cases and may cause issues in the application. 
So I replaced them with the username and email properties from the org.gitlab4j.api.models.User object to initialize the User superclass. 
This ensures that the GitLabOAuthUserDetails class is properly initialized with the correct values from the org.gitlab4j.api.models.User object.
correct me if I am wrong here.

Comment (MarkEWaite) [reply src/main/java/org/jenkinsci/plugins/GitLabOAuthUserDetails.java:15]: Changing the behavior of the system as part of a spotbugs warnings improvement is a dangerous choice.  I can't accept it.  I've harmed users more than once when I thought that a "cleanup" step was safe, but later proved that users relied on the previous behavior.

You're assuming that the empty string argument is worse than replacing it with another value, yet there is no evidence in any issue report or in any other location that the users want that change or that users will be unharmed by that change.  Authentication and authorization are critical parts of the system.  If this change inadvertently breaks a user, they could be locked out of their system.

If the change is needed, then it needs to happen in another pull request so that there will be a clearly phrased changelog entry that describes the behavior change to users.  The impact of the change needs to be assessed by reviewing how the caller uses the second argument and what impact the change from empty string to email address will have.

**Review by MarkEWaite — CHANGES_REQUESTED** [review]
The new call to `user.getEmail()` needs to be removed.  It does not belong in this pull request.

File: src/main/java/org/jenkinsci/plugins/GitLabOAuthUserDetails.java (hunk 5 above)
Comment (prince-panwar) [reply src/main/java/org/jenkinsci/plugins/GitLabOAuthUserDetails.java:15]: ok thanks for the explanation

**Review by MarkEWaite — APPROVED** [review]
Thanks again!
